packages/firestore/src/local/shared_client_state.ts

```diff
@@ -423,8 +571,81 @@ export class WebStorageSharedClientState implements SharedClientState {
    * Parses a client state key in LocalStorage. Returns null if the key does not
    * match the expected key format.
    */
-  private fromLocalStorageClientKey(key: string): string | null {
+  private fromLocalStorageClientStateKey(key: string): ClientKey | null {
     const match = this.clientStateKeyRe.exec(key);
     return match ? match[1] : null;
   }
+
+  /**
+   * Parses a client state in LocalStorage. Returns null if the key does not
+   * match the expected key format.
+   */
+  private fromLocalStorageClientState(
+    key: string,
+    value: string
+  ): RemoteClientState | null {
+    const clientId = this.fromLocalStorageClientStateKey(key);
+    return clientId != null
+      ? RemoteClientState.fromLocalStorageEntry(clientId, value)
+      : null;
+  }
+
+  /**
+   * Parses a mutation batch state in LocalStorage. Returns null if the key does
+   * not match the expected key format.
+   */
+  private fromLocalStorageMutation(
+    key: string,
+    value: string | null
+  ): RemoteMutationBatch | null {
+    const match = this.mutationBatchKeyRe.exec(key);
+
+    if (match) {
+      const isCurrentUser =
+        match[1] === 'null'
+          ? this.user.uid === null
+          : match[1] === this.user.uid;
```

Comment:
Hrm...  'null' could technically be a valid uid I think.  FWIW we just use an empty string for unauthenticated users in the mutation queue: https://github.com/firebase/firebase-js-sdk/blob/84896ec489ef162555c5b48881822982281d8921/packages/firestore/src/local/indexeddb_mutation_queue.ts#L83

Though having local storage keys with `__` to indicate unauthenticated is perhaps a bit awkward.

Oh, but on that note, I think `_` is allowed in uids and so our regex is going to be broken. :-/